src/Symfony/Component/Lock/Store/SemaphoreStore.php

```diff
@@ -102,7 +103,7 @@ public function delete(Key $key)
      */
     public function putOffExpiration(Key $key, $ttl)
     {
-        // do nothing, the semaphore locks forever.
+        throw new NotExpirableStoreException();
```

Comment:
Should be catch in `Lock::refresh`

Reply:
Are you sure? Because in the comment linked below I was suggesting it should be bubbled up in `Lock::refresh` as well because the client should not create a lock with `ttl`

https://github.com/symfony/symfony/issues/28696#issuecomment-427000248

Reply:
Okay - so `refresh` can be called outside of `save` as well since a lock can be refreshed at anytime. Yes, in that case I should catch and do nothing.

Reply:
yes, you're right. 

Issue is, services are not able to guess if the StoreInterface is able to extends  the TTL until calling the method and triggering an exception..

For instance, I could have a code running Semaphore in test/dev environement and redis in production and always calling `Lock->refesh()`...

I wonder if dealing with 2 interfaces wouldn't be easiest and simplest than this 2 exceptions...

```
interface ExpirableStoreInterface extends StoreInterface {
    public function putOffExpiration(Key $key, $ttl);
    public function saveIn(Key $key, $ttl);
}
class Lock {
  // either adding a new isRefreshable() or ignoring exception from not expirable stores...
  public function isRefreshable(): bool {return $this->store instanceOf ExpirableStoreInterface;}
}
```

In that case we have to duplicate/extends decoratorStore (CombinedStore and RetryTillSaveStore) in order to expose the right interface to the user
``` 
class CombinedStore implements StoreInterface {}
class ExpirableCombinedStore extends CombinedStore implements ExpirableStoreInterface {}

class RetryTillSaveStore implements StoreInterface {}
class ExpirableRetryTillSaveStore extends RetryTillSaveStore implements ExpirableStoreInterface {}
```
Or we can add a method `supportsExpiration` in that next interace. But it's weird
```
class ExpirableRetryTillSaveStore implements ExpirableStoreInterface {
  public function supportsExpiration(): bool {
    return $this->decorated instanceOf ExpirableStoreInterface && $this->decorated->supportsExpiration()
  }
}
```

Reply:
Your issue makes absolute sense. I also think it makes sense have a cleaner contract depending on what type of functionality the stores can support. The issue with leaving a noop in `putOffExpiration` will definitely cause the same issue you mentioned where if a user calls a `ttl` with `null` then an `InvalidArgumentException` is thrown.

Adding the new methods you suggested `isRefreshable` and the new `ExpirableStoreInterface` definitely makes the implementation more clean. I have updated the diff with the new design and remove the `NotExpirableStoreException` as it doesn't seem to be required anymore. 

I can't see any other obvious issues with this other than the fact that for `CombinedStores` we cannot combine an `Expirable Store` and `Non-expirable Store`. But I also don't see a good use case of having those two together in the decorator. 

Let me know your thoughts on this design.

Reply:
Based on https://github.com/symfony/symfony/pull/28727#discussion_r223501776
We are just going to catch and ignore the new created exception `NotExpirableStoreException`.

Reply:
Pinging @jderusse for review :)